### PR TITLE
nanotron <-> conversion for Llama resolve #124

### LINKED ISSUE #124: [Feature] nanotron <-> conversion for Llama
A similar idea as in https://github.com/huggingface/nanotron/pull/103, but for a Llama model.

I'd be happy to implement this.

I need it for another project that uses nanotron and was wondering if it is something that you'd want in this repository? If so, I'll start working on an implementation here.

Aside from the contribution guide, are there any other guidelines for this task? For example:
* Where should the conversion script be located?
* Any gotcha's I should be aware of?
* The best way to validate this would be to write a test that shows that the converted models return the same results as the non-converted ones. Do you think that a rather small model (that I can quickly iterate on while running locally) would be sufficient?

Thanks!

### PR DESCRIPTION
**Pull request by yardenas**
A similar idea as in https://github.com/huggingface/nanotron/pull/103, but for a Llama model.

I'd be happy to implement this.

I need it for another project that uses nanotron and was wondering if it is something that you'd want in this repository? If so, I'll start working on an implementation here.

Aside from the contribution guide, are there any other guidelines for this task? For example:

Where should the conversion script be located?
Any gotcha's I should be aware of?
The best way to validate this would be to write a test that shows that the converted models return the same results as the non-converted ones. Do you think that a rather small model (that I can quickly iterate on while running locally) would be sufficient?
Thanks!

### CODE REVIEW

**Comment by Lauler** [comment]
As an outsider also highly interested in seeing this conversion available in the library before I commit to train anything with Nanotron. Just here to cheer you on!

**Comment by xrsrke** [comment]
> Do you think that a rather small model (that I can quickly iterate on while running locally) would be sufficient?
Yup

**Comment by xrsrke** [comment]
Looks very nice. Please ping me once it's ready!!

**Comment by yardenas** [comment]
@xrsrke I think we're getting there. 
@AleHD made a significant progress making the tests actually pass -- we currently get ~0.02 absolute error on the logits in both directions.

wtyd?

**Comment by xrsrke** [comment]
@yardenas very cool. feel free to ping me if you need any pointers 🤗

**Comment by xrsrke** [comment]
> We made a copy of examples/doremi/tests/utils.py to the llama folder and made some modifications. It's testing utils for llama, so I wouldn't put it inside the nanotron library. That being said, it's not super DRY so happy to make changes if you think there's a better way

The test looks good 🙌

File: examples/llama/convert_hf_to_nanotron.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+"""
+Converts a HF model to nanotron format
+Command:
+    torchrun --nproc_per_node=1 convert_nanotron_to_hf.py --checkpoint_path=hf_weights --save_path=nanotron_weights
```
Comment (3outeille) [inline examples/llama/convert_hf_to_nanotron.py:4]: should be convert_hf_to_nanotron.py

File: examples/llama/tests/test_conversion.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
+# ruff: noqa: E402
+import json
+
+import pytest
+import torch
+from transformers import LlamaForCausalLM
+from utils import set_system_path
+
+set_system_path()
+
```
Comment (3outeille) [inline examples/llama/tests/test_conversion.py:10]: would be better to add `__init__.py` file in `tests/` and `examples/` + rename from `from convert_weights` to `from examples.llama.convert_weights` in both converter script

File: examples/llama/tests/utils.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import importlib
+import sys
+from pathlib import Path
+
+
+def set_system_path():
```
Comment (3outeille) [inline examples/llama/tests/utils.py:6]: won't be needed anymore with the `__init__.py`

File: examples/llama/tests/test_conversion.py (hunk 2 above)
Comment (yardenas) [reply examples/llama/tests/test_conversion.py]: @3outeille, done! looks good.

Question: we took this code snippet from the doremi example. Should we try to clean this up from there as well? I don't want to break it though

File: examples/llama/tests/utils.py (hunk 3 above)
Comment (3outeille) [reply examples/llama/tests/utils.py]: @yardenas okay I am very sorry, it was not a good idea as it break some stuff. Could you revert to your old way of doing please ?

File: examples/llama/convert_hf_to_nanotron.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    model_config = get_nt_config(hf_model.config)
+    nanotron_model = load_nanotron_model(model_config=model_config)
+
+    # Copy weights and save model.
+    parallel_context = nanotron.parallel.ParallelContext(
+        data_parallel_size=1, pipeline_parallel_size=1, tensor_parallel_size=1
+    )
+    convert_hf_to_nt(hf_model, nanotron_model, model_config)
+    nanotron.serialize.save_weights(model=nanotron_model, parallel_context=parallel_context, root_folder=save_path)
+    with open(save_path / "model_config.json", "w+") as f:
```
Comment (3outeille) [inline examples/llama/convert_hf_to_nanotron.py:108]: save config.yaml as well (to be able to run_generate in Nanotron side)

File: examples/llama/tests/utils.py (hunk 3 above)
Comment (yardenas) [reply examples/llama/tests/utils.py:6]: yes ofc. done

**Comment by yardenas** [comment]
> @yardenas Given a pretrained HF model (`huggyllama/llama-7b`), I tested the following:
> 
> * Convert HF to nanotron
> * Convert back Nanotron to HF
> * Run generate in HF (through `check_converted_model_generation()` in `convert_nanotron_to_hf.py`)
>   
>   * [x]  with cache
>   * [ ]  no cache (<== doesn't yield proper results)
> * Run generate in Nanotron
>   
>   * [ ]  with cache
>   * [x]  no cache

got it.
any ideas why it could fail for the no cache option?
I'll take a look into it

File: examples/llama/convert_hf_to_nanotron.py (hunk 4 above)
Comment (yardenas) [reply examples/llama/convert_hf_to_nanotron.py:108]: done.

Comment (3outeille) [reply examples/llama/convert_hf_to_nanotron.py:108]: Sorry I meant this:

```python
...
 parallel_config = nanotron.config.ParallelismArgs(
        dp=1,
        pp=1,
        tp=1,
        pp_engine=nanotron.config.AllForwardAllBackwardPipelineEngine(),
        tp_mode=nanotron.config.TensorParallelLinearMode.ALL_REDUCE,
        tp_linear_async_communication=False,
    )
    
    with open(save_path / "config.yaml", "w") as f:
        config = Config(
            general=GeneralArgs(project="test", run="llama"),
            parallelism=parallel_config,
            model=ModelArgs(
                init_method=RandomInit(std=0.2),
                model_config=model_config,
            ),
            tokenizer=TokenizerArgs(checkpoint_path), #tokenizer is inside the HF model
        )
        print("Saving config ...")
        yaml.dump(config.as_dict(), f)
   ...
  ```
  
  This is needed in https://github.com/huggingface/nanotron/blob/main/run_generate.py#L71

**Comment by 3outeille** [comment]
> > @yardenas Given a pretrained HF model (`huggyllama/llama-7b`), I tested the following:
> > 
> > * Convert HF to nanotron
> > * Convert back Nanotron to HF
> > * Run generate in HF (through `check_converted_model_generation()` in `convert_nanotron_to_hf.py`)
> >   
> >   * [x]   with cache
> >   * [ ]   no cache (<== doesn't yield proper results)
> > * Run generate in Nanotron
> >   
> >   * [ ]   with cache
> >   * [x]   no cache
> 
> got it. any ideas why it could fail for the no cache option? I'll take a look into it

Mhmm 

if I test it this way, generation is the same

```python
def check_converted_model_generation(save_path: Path):
    """Loads a huggingface model and tokenizer from `save_path` and
    performs a dummy text generation."""

    tokenizer = AutoTokenizer.from_pretrained(save_path)
    input_ids = tokenizer(TEST_PROMPT, return_tensors="pt")["input_ids"].cuda()
    print("Inputs:", tokenizer.batch_decode(input_ids))

    model = LlamaForCausalLM.from_pretrained(save_path).cuda().bfloat16()
    out = model.generate(input_ids, max_new_tokens=100)
    print("Generation (converted): ", tokenizer.batch_decode(out))
    
    model_nocache = LlamaForCausalLM.from_pretrained(save_path).cuda().bfloat16()
    model_nocache.config.use_cache = False
    out_nocache = model_nocache.generate(input_ids, max_new_tokens=100)
    print("Generation (converted, no cache): ", tokenizer.batch_decode(out_nocache))
 ```

However, if I do step `Convert back Nanotron to HF`  then  manually change `use_cache=False` in `model_config.json` before step `Run generate in HF (through check_converted_model_generation() in convert_nanotron_to_hf.py)` , then the generation is not good (maybe I am not supposed to do that)

**Comment by 3outeille** [comment]
@yardenas Also, have you tried training a llama in Nanotron in `DP=PP=1 & TP=2`and run `convert_nanotron_to_hf.py` ?

File: examples/llama/convert_hf_to_nanotron.py (hunk 4 above)
Comment (yardenas) [reply examples/llama/convert_hf_to_nanotron.py:109]: fixed. thanks!

btw, I would suggest making a followup PR where parallelism parameters can be defined by users (instead of bluntly hardcoding them). wdyt?

File: examples/llama/convert_nanotron_to_hf.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    a new huggingface instance, copies the weights from the nanotron checkpoint
+    and saves the transformed huggingface to `save_path`."""
+
+    # Init nanotron model.
+    device = torch.device("cuda")
+    with open(checkpoint_path / "model_config.json", "r") as f:
+        attrs = json.load(f)
+        model_config = NanotronLlamaConfig(**attrs)
+    dtype = getattr(torch, "bfloat16")
+    nanotron_model = load_nanotron_model(
```
Comment (3outeille) [inline examples/llama/convert_nanotron_to_hf.py:112]: you forced tp/dp/pp = 1 here. If someone trained a nanotron model in TP=2 for example, he will not be able to convert to HF as it forces tp=1

Comment (3outeille) [inline examples/llama/convert_nanotron_to_hf.py:112]: you forced tp/dp/pp = 1 here. If someone trained a nanotron model in TP=2 for example, he will not be able to convert to HF as it forces tp=1

File: examples/llama/tests/test_conversion.py [hunk 6]
```diff
@@ -73,120 +67,136 @@ def input_ids() -> torch.Tensor:
...
+def _test_tensor_parallel_conversion(parallel_context: ParallelContext):
+    # model_nt = create_nanotron_model(tp=2)
+    # model_hf = create_huggingface_model()
+    # convert_nt_to_hf(model_nt, model_hf, CONFIG)
+    # input_mask = torch.ones_like(input_ids)
+    # logits_nt = model_nt.model(input_ids, input_mask).permute(1, 0, 2)
+    # logits_hf = model_hf(input_ids).logits
+    # assert logits_nt.size() == logits_hf.size()
+    # assert torch.allclose(logits_nt, logits_hf, atol=ATOL), torch.mean(torch.abs(logits_nt - logits_hf))
+    assert True
```
Comment (3outeille) [inline examples/llama/tests/test_conversion.py:198]: is there a reason why all tests were commented out ?

File: examples/llama/convert_nanotron_to_hf.py (hunk 5 above)
Comment (yardenas) [reply examples/llama/convert_nanotron_to_hf.py:118]: @3outeille, added code that loads the `config.yaml` file and reads the parallelism parameters from there.

btw, I tried running `test_tensor_parallel_conversion", but for some reason `init_distributed` hangs in `mp.spawn(...)`.

So I couldn't really test that it's working (and kinda stuck). Any idea why this can happen?

Comment (3outeille) [reply examples/llama/convert_nanotron_to_hf.py:118]: sorry for the delay, will take a look

**Comment by yardenas** [comment]
@3outeille we added a fix for the tp=2 case. :innocent:

File: examples/llama/tests/test_conversion.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    input_mask = torch.ones_like(input_ids)
+    logits_nt = model_nt.model(input_ids, input_mask).permute(1, 0, 2)
+    del model_nt
+    # Perform conversion.
+    convert_nt_to_hf_and_save(nt_path, hf_path)
+    # Load huggingface and get logits.
+    model_hf = LlamaForCausalLM.from_pretrained(hf_path).cuda()
+    logits_hf = model_hf(input_ids).logits
+    assert logits_nt.size() == logits_hf.size()
+    assert torch.allclose(logits_nt, logits_hf, atol=ATOL), torch.mean(torch.abs(logits_nt - logits_hf))
```
Comment (xrsrke) [inline examples/llama/tests/test_conversion.py:129]:
```suggestion
    torch.testing.assert_allclose(logits_nt, logits_hf, atol=ATOL)
```

File: examples/llama/tests/test_conversion.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+def _test_hf_to_nt(parallel_context: ParallelContext, input_ids: torch.Tensor):
+    model_nt = create_nanotron_model(parallel_context)
+    model_hf = create_huggingface_model()
+    convert_hf_to_nt(model_hf, model_nt, CONFIG)
+    input_mask = torch.ones_like(input_ids)
+    logits_nt = model_nt.model(input_ids, input_mask).permute(1, 0, 2)
+    logits_hf = model_hf(input_ids).logits
+    assert logits_nt.size() == logits_hf.size()
+    assert torch.allclose(logits_nt, logits_hf, atol=ATOL), torch.mean(torch.abs(logits_nt - logits_hf))
```
Comment (xrsrke) [inline examples/llama/tests/test_conversion.py:144]:
```suggestion
    torch.testing.assert_allclose(logits_hf, logits_nt, atol=ATOL)  
```

File: examples/llama/tests/test_conversion.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+def _convert_from_parallel(parallel_context: ParallelContext, input_ids: torch.Tensor, nt_path: Path, hf_path: Path):
+    # Convert parallel nanotron to hf, get and save huggingface predictions.
+    convert_nt_to_hf_and_save(nt_path, hf_path)
+    model_hf = LlamaForCausalLM.from_pretrained(hf_path).cuda()
+    logits_hf = model_hf(input_ids).logits
+    torch.save(logits_hf.detach().cpu(), hf_path / "logits.pt")
+
+
+def test_tensor_parallel_conversion(input_ids: torch.Tensor):
```
Comment (xrsrke) [inline examples/llama/tests/test_conversion.py:234]: this test is the same as test_composed_conversion?

File: examples/llama/tests/test_conversion.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    # Convert to nanotron one more time, compare statedicts.
+    del model_nt
+    model_nt = create_nanotron_model(parallel_context)
+    convert_hf_to_nt(model_hf, model_nt, CONFIG)
+    nt_sd_new = model_nt.state_dict()
+    assert set(nt_sd_new) == set(nt_sd)
+    assert all(torch.all(nt_sd[key] == nt_sd_new[key]) for key in nt_sd_new)
+
+
+def test_composed_conversion():
```
Comment (xrsrke) [inline examples/llama/tests/test_conversion.py:198]: very nice test 🙌 !!

Comment (yardenas) [reply examples/llama/tests/test_conversion.py:198]: @xrsrke thank you! (it was mainly written by @AleHD!)

**Comment by yardenas** [comment]
@xrsrke and @3outeille I just committed the requests from @xrsrke, anything else? :)
